src/main/java/com/stuypulse/robot/subsystems/Shooter.java

```diff
@@ -5,24 +5,152 @@
+
+        feederPIDController.setP(FEEDER_P);
+        feederPIDController.setI(FEEDER_I);
+        feederPIDController.setD(FEEDER_D);
+        feederPIDController.setFF(FEEDER_FF);
+
+        // put the PID values for shooter and feder onto the Smartdashboard
+        SmartDashboard.putNumber("Shooter P value" , SHOOTER_P);
+        SmartDashboard.putNumber("Shooter I value" , SHOOTER_I);
+        SmartDashboard.putNumber("Shooter D value" , SHOOTER_D);
+        SmartDashboard.putNumber("Shooter F value" , SHOOTER_FF);
+
+        SmartDashboard.putNumber("Feeder P value" , FEEDER_P);
+        SmartDashboard.putNumber("Feeder I value" , FEEDER_I);
+        SmartDashboard.putNumber("Feeder D value" , FEEDER_D);
+        SmartDashboard.putNumber("Feeder F value" , FEEDER_FF);
+
+
+    }
+
+    public void setShooterRPM(double speed) {
+        shooterMotor.set(speed);
+        feederMotor.set(speed);
+    }
+
+    public double getShooterRPM() {
+        // units are RPM
+        return (Math.abs(encoderA.getVelocity())
+        + Math.abs(encoderB.getVelocity())) / 2;
+    }
+
+    public double getFeederRPM() {
+        // units are RPM
+        return Math.abs(feederEncoder.getVelocity());
     }
 
     @Override
     public void periodic() {
         // This method will be called once per scheduler run
+        // for the ring shot
+        
+
+        // get the current value for p, i, and d from smart dashboard
+        shooter_P = SmartDashboard.getNumber("Shooter P value" , SHOOTER_P);
+        shooter_I = SmartDashboard.getNumber("Shooter I value" , SHOOTER_I);
+        shooter_D = SmartDashboard.getNumber("Shooter D value" , SHOOTER_D);
+        shooter_F = SmartDashboard.getNumber("Shooter F value" , SHOOTER_FF);
+        
+        feeder_P = SmartDashboard.getNumber("Feeder P value" , FEEDER_P);
+        feeder_I = SmartDashboard.getNumber("Feeder I value" , FEEDER_I);
+        feeder_D = SmartDashboard.getNumber("Feeder D value" , FEEDER_D);
+        feeder_F = SmartDashboard.getNumber("Feeder F value" , FEEDER_FF);
+
+        // If the value for PID changed on smart dashboard, change it here too
+        if((shooter_P != SHOOTER_P)) { shooterPIDController.setP(shooter_P); SHOOTER_P = shooter_P; }
+        if((shooter_I != SHOOTER_I)) { shooterPIDController.setP(shooter_I); SHOOTER_I = shooter_I; }
+        if((shooter_D != SHOOTER_D)) { shooterPIDController.setP(shooter_D); SHOOTER_D = shooter_D; }
+        if((shooter_D != SHOOTER_D)) { shooterPIDController.setP(shooter_D); SHOOTER_FF = shooter_F; }
+
+        if((shooter_P != FEEDER_P)) { shooterPIDController.setP(shooter_P); FEEDER_P = shooter_P; }
+        if((shooter_I != FEEDER_I)) { shooterPIDController.setP(shooter_I); FEEDER_I = shooter_I; }
+        if((shooter_D != FEEDER_D)) { shooterPIDController.setP(shooter_D); FEEDER_D = shooter_D; }
+        if((shooter_D != FEEDER_D)) { shooterPIDController.setP(shooter_D); FEEDER_FF = shooter_F; }
```

Comment:
There are many typos in this block of code.

- You repeat D multiple times
- you only ever set Shooter P
- You don't actually change feeder values
- You change constant values